statsd/pipe_windows.go

```diff
@@ -3,20 +3,82 @@
 package statsd
 
 import (
-	"errors"
 	"net"
+	"sync"
 	"time"
 
 	"github.com/Microsoft/go-winio"
 )
 
-type pipeWriter struct{ net.Conn }
+const defaultPipeTimeout = 1 * time.Millisecond
 
-func (pipeWriter) SetWriteTimeout(_ time.Duration) error {
-	return errors.New("SetWriteTimeout is not supported on Windows Named Pipes")
+type pipeWriter struct {
+	mu       sync.RWMutex
+	conn     net.Conn
+	timeout  time.Duration
+	pipepath string
+}
+
+func (p *pipeWriter) SetWriteTimeout(d time.Duration) error {
+	p.mu.Lock()
+	p.timeout = d
+	p.mu.Unlock()
+	return nil
+}
+
+func (p *pipeWriter) Write(data []byte) (n int, err error) {
+	conn, err := p.ensureConnection()
+	if err != nil {
+		return 0, err
+	}
+
+	p.mu.RLock()
+	conn.SetWriteDeadline(time.Now().Add(p.timeout))
+	p.mu.RUnlock()
+
+	n, err = conn.Write(data)
```

Comment:
What happens if 2 go routines call `conn.Write(data)` at the same time. Is the server able to read the first data and then the second data or the data from call 1 and 2 are mixed?

Reply:
I don't know. The same question applies to the UDS implementation I guess. The only way out of it is introducing a mutex. Do we want that? I'd think no.

Reply:
As far as I remember, UDS does not have this issue as it uses datagram.

Reply:
I wrote a test to try this out. It creates 100 goroutines each sending 100 metrics simultaneously and then checks the output confirming that each individual metric is there. It seems fine, even when running the tests many times in a row (with and without `-race` flag and with `-count=20`). Code is here: https://github.com/DataDog/datadog-go/compare/gbbr/test-concur?expand=1